### PR TITLE
Write a termination monitor for spot instances

### PR DESCRIPTION
**Pull request by KJTsanaktsidis**
This puppy will detect when it is about to be terminated, and reschedule
jobs.

### CODE REVIEW

File: packer/conf/bin/spot_termination_monitor.rb [hunk 1]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env ruby
+
+require 'httparty'
+require 'socket'
+
+# Loop and poll the metadata route for spot instance termination
+# When we are terminated...
+#     1) Use hacks to print a message to every buildkite output
```
Comment (rianmcguire) [inline packer/conf/bin/spot_termination_monitor.rb:8]: Are you about to do what I think you are?

File: packer/conf/bin/spot_termination_monitor.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        pid = File.basename(proc).to_i
+        puts "Sending SIGQUIT to #{pid}"
+        Process.kill('QUIT', pid)
+    end
+
+    # Give everything a few seconds to do their thing - the effect of the SIGQUIT is not synchronous
+    while Dir.glob('/proc/[0-9]*').any? { |proc| File.readlink(File.join(proc, 'exe')).start_with?('/usr/bin/buildkite-agent') rescue false }
+        puts "Waiting for buildkite-agent processes to exit..."
+        sleep 1
+    end
```
Comment (rianmcguire) [inline packer/conf/bin/spot_termination_monitor.rb:74]: Nothing's going to restart this automatically?

File: packer/conf/bin/spot_termination_monitor.rb (hunk 1 above)
Comment (rianmcguire) [reply packer/conf/bin/spot_termination_monitor.rb:8]: Ewww

**Comment by rianmcguire** [comment]
Totally gross...but happy to 🚢 if you are.

File: packer/conf/bin/spot_termination_monitor.rb (hunk 2 above)
Comment (KJTsanaktsidis) [reply packer/conf/bin/spot_termination_monitor.rb:74]: Nope.

Comment (KJTsanaktsidis) [reply packer/conf/bin/spot_termination_monitor.rb:74]: Sysvinit, you think it has process monitoring? :P

**Comment by KJTsanaktsidis** [comment]
FYI a previous version of this looked at `/proc/pid/cmdline` but i figured this was more robust.

**Pull request merged**